x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/state.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient, Logger } from '@kbn/core/server';
+import { StreamsStorageClient } from '../service';
+import { StreamActiveRecord, StreamChange, ValidationResult } from './types';
+import { streamFromDefinition } from './stream_from_definition';
+
+export class State {
+  private streamsByName: Map<string, StreamActiveRecord>;
+
+  constructor(streams: StreamActiveRecord[]) {
+    this.streamsByName = new Map();
+    streams.forEach((stream) => this.streamsByName.set(stream.definition.name, stream));
+  }
+
+  applyChanges(requestedChanges: StreamChange[]): State {
+    // What if multiple changes target the same stream?
+
+    // Optional: Expand requested changes to include automatic creation of missing streams
+    const newState = this.clone();
+
+    requestedChanges.forEach((change) => {
+      const targetStream = newState.get(change.target);
+
+      if (!targetStream) {
+        if (change.type === 'delete') {
+          throw new Error('Cannot delete non-existing stream');
+        } else {
+          const newStream = streamFromDefinition(change.request.stream);
+          newStream.update(change.request.stream, newState, this);
```

Comment:
This is a little bit of a "hack", but this allows the newly created stream to a) be marked as changed and b) gives it a chance to affect any other streams as part of its creation